### ISSUE TITLE
BLIP caption finetune on custom dataset

### ISSUE THREAD

**Issue by gunesevitan** [salesforce/LAVIS#218]
I couldn't find an example about this task so I integrated default model behavior into my training loop. As one of the model output is loss, I can directly step on that and update model weights. Does that work seamlessly or am I missing anything else? Here is my training function.

```
def train(train_loader, model, optimizer, device, scheduler=None, amp=False):

    """
    Train given model on given data loader

    Parameters
    ----------
    train_loader: torch.utils.data.DataLoader
        Training set data loader

    model: torch.nn.Module
        Model to train

    optimizer: torch.optim.Optimizer
        Optimizer

    device: torch.device
        Location of the model and inputs

    scheduler: torch.optim.LRScheduler or None
        Learning rate scheduler

    amp: bool
        Whether to use auto mixed precision or not

    Returns
    -------
    running_train_loss: float
        Running training loss after model is fully trained on training set data loader
    """

    model.train()
    progress_bar = tqdm(train_loader)
    running_train_loss = 0.0

    if amp:
        grad_scaler = torch.cuda.amp.GradScaler()
    else:
        grad_scaler = None

    for step, (images, captions) in enumerate(progress_bar):

        images = images.to(device)
        optimizer.zero_grad()

        if amp:
            with torch.cuda.amp.autocast():
                outputs = model(samples={
                    'image': images,
                    'text_input': captions
                })

            loss = outputs['loss']
            grad_scaler.scale(loss).backward()
            grad_scaler.step(optimizer)
            grad_scaler.update()
        else:
            outputs = model(samples={
                'image': images,
                'text_input': captions
            })
            loss = outputs['loss']
            loss.backward()
            optimizer.step()

        if scheduler is not None:
            scheduler.step()

        running_train_loss += loss.detach().item() * images.size(0)
        lr = scheduler.get_last_lr()[0] if scheduler is not None else optimizer.param_groups[0]['lr']
        progress_bar.set_description(f'train loss: {running_train_loss / len(train_loader.sampler):.6f} - lr: {lr:.8f}')

    running_train_loss = running_train_loss / len(train_loader.sampler)

    return running_train_loss
```

**Comment by dxli94** [comment]
It looks good to me.

We actually have a tutorial that you may take a look as reference https://opensource.salesforce.com/LAVIS//latest/tutorial.training-example.html
